azure/services/privateendpoints/spec.go

```diff
@@ -54,209 +51,118 @@ type PrivateEndpointSpec struct {
-func (s *PrivateEndpointSpec) ResourceName() string {
-	return s.Name
-}
-
-// ResourceGroupName returns the name of the resource group.
-func (s *PrivateEndpointSpec) ResourceGroupName() string {
-	return s.ResourceGroup
-}
-
-// OwnerResourceName is a no-op for private endpoints.
-func (s *PrivateEndpointSpec) OwnerResourceName() string {
-	return ""
-}
-
-// Parameters returns the parameters for the PrivateEndpointSpec.
-func (s *PrivateEndpointSpec) Parameters(ctx context.Context, existing interface{}) (interface{}, error) {
-	_, log, done := tele.StartSpanWithLogger(ctx, "privateendpoints.Service.Parameters")
-	defer done()
-
-	privateEndpointProperties := armnetwork.PrivateEndpointProperties{
-		Subnet: &armnetwork.Subnet{
-			ID: &s.SubnetID,
-			Properties: &armnetwork.SubnetPropertiesFormat{
-				PrivateEndpointNetworkPolicies:    ptr.To(armnetwork.VirtualNetworkPrivateEndpointNetworkPoliciesDisabled),
-				PrivateLinkServiceNetworkPolicies: ptr.To(armnetwork.VirtualNetworkPrivateLinkServiceNetworkPoliciesEnabled),
-			},
+// ResourceRef implements azure.ASOResourceSpecGetter.
+func (s *PrivateEndpointSpec) ResourceRef() *asonetworkv1.PrivateEndpoint {
+	return &asonetworkv1.PrivateEndpoint{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      s.Name,
+			Namespace: s.Namespace,
 		},
 	}
+}
 
-	privateEndpointProperties.CustomNetworkInterfaceName = ptr.To(s.CustomNetworkInterfaceName)
-
-	privateIPAddresses := make([]*armnetwork.PrivateEndpointIPConfiguration, 0, len(s.PrivateIPAddresses))
-	for _, address := range s.PrivateIPAddresses {
-		ipConfig := &armnetwork.PrivateEndpointIPConfigurationProperties{PrivateIPAddress: ptr.To(address)}
-
-		privateIPAddresses = append(privateIPAddresses, &armnetwork.PrivateEndpointIPConfiguration{
-			Properties: ipConfig,
+// Parameters implements azure.ASOResourceSpecGetter.
+func (s *PrivateEndpointSpec) Parameters(ctx context.Context, existingPrivateEndpoint *asonetworkv1.PrivateEndpoint) (*asonetworkv1.PrivateEndpoint, error) {
+	privateEndpoint := &asonetworkv1.PrivateEndpoint{}
+	privateEndpoint.Spec = asonetworkv1.PrivateEndpoint_Spec{}
+	if existingPrivateEndpoint != nil {
+		privateEndpoint = existingPrivateEndpoint
+	}
+
+	// ApplicationSecurityGroups
+	if len(s.ApplicationSecurityGroups) > 0 {
+		applicationSecurityGroups := make([]asonetworkv1.ApplicationSecurityGroupSpec_PrivateEndpoint_SubResourceEmbedded, 0, len(s.ApplicationSecurityGroups))
+		for _, applicationSecurityGroup := range s.ApplicationSecurityGroups {
+			applicationSecurityGroups = append(applicationSecurityGroups, asonetworkv1.ApplicationSecurityGroupSpec_PrivateEndpoint_SubResourceEmbedded{
+				Reference: &genruntime.ResourceReference{
+					ARMID: applicationSecurityGroup,
+				},
+			})
+		}
+		sort.SliceStable(applicationSecurityGroups, func(i, j int) bool {
+			return applicationSecurityGroups[i].Reference.ARMID < applicationSecurityGroups[j].Reference.ARMID
 		})
```

Comment:
Why is this sort needed? I think a comment explaining this could be helpful.

Reply:
This sort was carried over from the previous code flow. 
Earlier code seems to be preserving the order of both new and existing private endpoints. xref: https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/4108/files#diff-c5a1236250878b000a50b1ce5a20dd4a7412176fd3e8bcb34f48e7dab6ac8a7cL163-L165 
But, yes, adding a comment is a good idea, will update.